spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,89 +113,251 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
-						}
-						registerHandlerMethodOnListenedChannel(method, streamListener, bean);
+					int methodArgumentsLength = method.getParameterTypes().length;
+					if (!isDeclarative) {
+						assertStreamListenerMessageHandlerMethod(method, methodAnnotatedInboundName);
+						registerHandlerMethodOnListenedChannel(method, methodAnnotatedInboundName, methodAnnotatedOutboundName, bean);
 					}
 					else {
-						for (int i = 0; i < parameterTypes.length; i++) {
-							MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, i);
-							Assert.isTrue(methodParameter.getParameterAnnotation(Input.class) != null ^
-											methodParameter.getParameterAnnotation(Output.class) != null,
-									"A declarative @StreamListener method must have its parameters annotated" +
-											"with @Input or @Output, but not with both.");
-						}
-						if (!method.getReturnType().equals(Void.TYPE)) {
-							Assert.isTrue(method.getAnnotation(Output.class) != null,
-									"A declarative @StreamListener method must be annotated with @Output");
+						Object[] declarativeArguments = new Object[methodArgumentsLength];
+						for (int parameterIndex = 0; parameterIndex < methodArgumentsLength; parameterIndex++) {
+							MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+							Object targetBean = null;
+							if (methodParameter.hasParameterAnnotation(Input.class)) {
+								String inboundElementName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Input.class));
+								Assert.isTrue(StringUtils.hasText(inboundElementName), "@Input annotation should always be associated with a valid inbound name");
+								targetBean = getBindableBean(inboundElementName);
+								declarativeArguments[parameterIndex] = getDeclarativeArgument(targetBean, methodParameter);
+							}
+							else if (methodParameter.hasParameterAnnotation(Output.class)) {
+								String outboundElementName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Output.class));
+								Assert.isTrue(StringUtils.hasText(outboundElementName), "@Output annotation should always be associated with a valid outbound name");
+								targetBean = getBindableBean(outboundElementName);
+								declarativeArguments[parameterIndex] = getDeclarativeArgument(targetBean, methodParameter);
+							}
+							else if (declarativeArguments.length == 1 && StringUtils.hasText(methodAnnotatedInboundName)) {
+								targetBean = getBindableBean(methodAnnotatedInboundName);
+								declarativeArguments[parameterIndex] = getDeclarativeArgument(targetBean, methodParameter);
+							}
+							Assert.isTrue(declarativeArguments[parameterIndex] != null, "Declarative StreamListener method should only have " +
+									"inbound or outbound targets as method parameters");
+							Assert.notNull(declarativeArguments[parameterIndex],
```

Comment:
this would never be invoked given that the assertion on the previous line would have already failed